Documentation/proposals/202305-scrapeclasses.md

```diff
@@ -0,0 +1,189 @@
+    path: /metrics
+    scrapeClass: istio-mtls
+```
+
+The suggested behavior is:
+1. the `tlsConfig` in the associated scrape class is automatically applied to the scrape configuration of the endpoint.
+2. the inline `tlsConfig` (if any) takes precedence over the `tlsConfig` in the scrape class.
+
+### Probe Resource
+Allow the user to select a scrape class for the probe.
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: Probe
+spec:
+  scrapeClass: istio-mtls
+```
+
+### ServiceMonitor Resource
+Allow the user to select a scrape class for each endpoint.
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+spec:
+  endpoints:
+  - port: http
+    path: /metrics
+    scrapeClass: istio-mtls
+```
+
+Deprecate the unsafe TLS settings in `ServiceMonitor`.
+
+## Test Plan
+
+1. Regression test; ensure no change to generated configs unless a scrape class is applied.
+2. Optionality of the default scrape class; ensure that it is an optional configuration element.
+3. Acceptance testing in Istio environment; ensure that the solution is effective for a key use case.
+
+## Alternatives
+
+### Global scrape TLS configuration
+An alternative solution would be to apply a default TLS configuration to all monitors.
+
+For example, via a hypothetical field `spec.scrapeTlsConfig`:
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: Prometheus
+spec:
+  scrapeTlsConfig:
+    caFile: "/etc/istio-certs/root-cert.pem"
+    certFile: "/etc/istio-certs/cert-chain.pem"
+    keyFile: "/etc/istio-certs/key.pem"
+    insecureSkipVerify: true
+```
+
+Objections:
+1. A singular default configuration may be too inflexible to effectively scrape a diverse set of pods. For example, in Istio, 
+some pods may be in STRICT mode.
+
+### Istio Permissive Mode
+An alternative for the Istio use case is to use PERMISSIVE mode (see [documentation](https://istio.io/latest/docs/concepts/security/#permissive-mode)),
+or to use `exclude` annotations on the pod such that the metrics endpoint bypasses mTLS.
+
+Objections:
+1. This solution is less secure at a transport level unless TLS is implemented at the application layer.
+2. Strict mTLS is the basis for [Istio authorization policy](https://istio.io/latest/docs/reference/config/security/authorization-policy/), e.g. to explicitly allow access from prom to the metrics endpoint.
+
+### ScrapeClass Resource
+A variant of the proposed solution is to introduce a new custom resource for defining scrape classes.
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: ScrapeClass
+metadata:
+  name: istio
+spec:
+  tlsConfig:
+    caFile: "/etc/istio-certs/root-cert.pem"
+    certFile: "/etc/istio-certs/cert-chain.pem"
+    keyFile: "/etc/istio-certs/key.pem"
+    insecureSkipVerify: true
+```
+
+An open question is whether the resource would be cluster-scoped or namespace-scoped.
+
+Objections:
+1. Since the file paths are dependent on the volume mounts in the server, this approach may not achieve a meaningful decoupling.
```

Comment:
should the `ScrapeClass` be limited to just the `tlsConfig` and its files though? I could imagine it holding the `ProberSpec` as well for instance, which could be more dynamic than the Prometheus deployment.

Reply:
not sure I'm following.

Reply:
Linking to my other comment for the complete context: https://github.com/prometheus-operator/prometheus-operator/pull/5572#discussion_r1192333598

Reply:
Just giving my .02 cents I agree with your use case on https://github.com/prometheus-operator/prometheus-operator/pull/5572#discussion_r1192333598 but I'm still not 100% convinced.

As of now the main advantages to SrapeClass being their own CRD that I'm able to see are:
1. Users being able to do `kc get ScrapeClass`  to see the available classes and pick the more desirable one
2. If there are many ScrapeClasses then maintaining a big list in the `Prometheus`  resource might become cumbersome

Reply:
The only other point I can see is to decouple the Prometheus CRD from the ScrapeClasses in the event of a Probe-like target managed by a different team (where ScrapeClasses could be managed by that same team). Now this might be a use case that's too little to consider.

After reading @EronWright's message on #prometheus-operator-dev (https://kubernetes.slack.com/archives/C01B03QCSMN/p1684516562355139?thread_ts=1684439235.731889&cid=C01B03QCSMN), I think we should be fine starting with inline, and can, if needed, reconsider a CRD at a later stage.
> 1. the basic idea is to separate the administrative inputs from the user inputs, and it seems likely to go beyond TLS, e.g. file-based bearer tokens.
> 2. Inline vs CRD aren’t mutually exclusive, and given the dependency on file mounts, I have a mild preference for inline class definition as a starting point.  If we go for a CRD, then please consider whether it would be namespaced and which namespace (prom or monitor?).
> 3. Is “scrape class” a good name, or would “monitor class” be better?  Are probes a type of monitor?

Reply:
Given the added complexity of introducing a `ScrapeClass` CR, I propose that we start with inline classes (marked as experimental) and revisit later.